### PR TITLE
Simplify type check in ConditionalProjection

### LINKED ISSUE #1180: NH-3847 - ConditionalProjection throws "Both true and false projections must return the same types" when the types are the same
**Alan Carroll** created an [issue](https://nhibernate.jira.com/browse/NH-3847) — <time datetime="2016-02-18T15:50:29.063Z" title="2016-02-18T15:50:29.063Z">18th February 2016, 15:50:29</time>:
> The code below generates the exception shown. Client, company, and vendor name are all AnsiString type, with a length of 200.  The problem stems from the type equality comparison, which uses the string length. `NHibernateUtil.AnsiString` does not have a length defined. Perhaps exact type equality is not what we want here.
> 
> 
> ```
> 
> var associate = Projections.Conditional(
>                 Restrictions.IsNotNull(Projections.Property(() => this.root.Client)),   Projections.Property(() => this.client.Name),   Projections.Conditional(
>                 Restrictions.IsNotNull(Projections.Property(() => this.root.Company)),  Projections.Property(() => this.company.Name),  Projections.Conditional(
>                 Restrictions.IsNotNull(Projections.Property(() => this.root.Vendor)),   Projections.Property(() => this.vendor.Name),   Projections.Constant(null, NHibernateUtil.AnsiString))));
> 
> ```
> 
> 
> {noformat:title=Exception}
> Both true and false projections must return the same types.
> But True projection returns: <NHibernate.Type.AnsiStringType> 
> And False projection returns: [NHibernate.Type.AnsiStringType]
> ```
> 
> {noformat:title=Stack Trace}
> [HibernateException: Both true and false projections must return the same types.
> But True projection returns: [NHibernate.Type.AnsiStringType] 
> And False projection returns: [NHibernate.Type.AnsiStringType]]
>    NHibernate.Criterion.ConditionalProjection.GetTypes(ICriteria criteria, ICriteriaQuery criteriaQuery)
>    NHibernate.Criterion.SimpleProjection.GetColumnCount(ICriteria criteria, ICriteriaQuery criteriaQuery)
>    NHibernate.Criterion.SimpleProjection.GetColumnAliases(Int32 position, ICriteria criteria, ICriteriaQuery criteriaQuery)
>    NHibernate.Criterion.ConditionalProjection.ToSqlString(ICriteria criteria, Int32 position, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    NHibernate.Criterion.ConditionalProjection.ToSqlString(ICriteria criteria, Int32 position, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    NHibernate.Criterion.ConditionalProjection.ToSqlString(ICriteria criteria, Int32 position, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    NHibernate.Criterion.AliasedProjection.ToSqlString(ICriteria criteria, Int32 position, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    NHibernate.Criterion.ProjectionList.ToSqlString(ICriteria criteria, Int32 loc, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    NHibernate.Loader.Criteria.CriteriaQueryTranslator.GetSelect(IDictionary`2 enabledFilters)
>    NHibernate.Loader.Criteria.CriteriaJoinWalker..ctor(IOuterJoinLoadable persister, CriteriaQueryTranslator translator, ISessionFactoryImplementor factory, ICriteria criteria, String rootEntityName, IDictionary`2 enabledFilters)
>    NHibernate.Loader.Criteria.CriteriaLoader..ctor(IOuterJoinLoadable persister, ISessionFactoryImplementor factory, CriteriaImpl rootCriteria, String rootEntityName, IDictionary`2 enabledFilters)
>    NHibernate.Impl.SessionImpl.List(CriteriaImpl criteria, IList results)
>    NHibernate.Impl.CriteriaImpl.List(IList results)
>    NHibernate.Impl.CriteriaImpl.List<T>()
>    NHibernate.Criterion.QueryOver`1.List<U>()
>    NHibernate.Criterion.QueryOver`1.NHibernate.IQueryOver<TRoot>.List<U>()
> ```
> 
> Since AnsiStringType does not have a public constructor, we cannot create our own with a matching length.  My temporary workaround is to create my own class that inherits from AbstractStringType, and specify the length.
> 
> 
> ```
> 
> private class MyAnsiStringType : NHibernate.Type.AbstractStringType
>         {
>             public MyAnsiStringType() : base (new NHibernate.SqlTypes.AnsiStringSqlType(200)) { }
> 
>             public override string Name
>             {
>                 get { return "AnsiString";  }
>             }
>         }
> 
> ```
> 
> 
> This works, since the lengths are all the same, but I should be able to do a conditional that returns strings of varying lengths. Furthermore, I should be able to return any string type, because this is valid SQL (SQL Server at least).
> 
> 
> ```sql
> 
> select
> 	case
> 		when 1 = 1 then '1'
> 		when 1 = 2 then N'2'
> 	end
> 
> ```
> 
> 
---
**Alan Carroll** added a comment — <time datetime="2017-06-14T16:54:36.393Z" title="2017-06-14T16:54:36.393Z">14th June 2017, 16:54:36</time>:
> I've found a simpler way to work around this issue using the built-in `NHibernate.Type.TypeFactory`.
> 
> 
> ```
> TypeFactory.GetAnsiStringType(200)
> ```
> 
>

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #1180

### CODE REVIEW

File: src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@
 namespace NHibernate.Test.NHSpecificTest.GH1180
 {
 	using System.Threading.Tasks;
-	[KnownBug("NH-3847 (GH-1180)")]
+	//NH-3847 (GH-1180)
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs:19]: Dont need this commet. It is alterady in that namespace

**Review by hazzik — APPROVED** [review]

File: src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs [hunk 2]
```diff
@@ -100,46 +101,47 @@ public async Task DecimalTypesAsync()
...
+					//Property scale is 2, make sure constant scale 3 is not lost
+					Projections.Constant(20.123m),
 					Projections.Property(nameof(Entity.Amount)));
 				tagCriteria.SetProjection(conditionalProjection);
 
 				// run query
 				var results = await (tagCriteria.ListAsync());
 
-				Assert.That(results, Is.EquivalentTo(new[] {20m, 42.13m, 20m}));
+				Assert.That(results, Is.EquivalentTo(new[] {20.123m, 42.13m, 20.123m}));
```
Comment (bahusoid) [inline src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs:127]: Is that expected that SQLite saves Scale more than defined? I expected 42.131m to be saved as 42.13m according to mapping Scale value 2.

File: src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs (hunk 1 above)
Comment (bahusoid) [reply src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs]: It's been added to mention NH-3847... Ok GH part deleted

File: src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs (hunk 2 above)
Comment (fredericDelaporte) [reply src/NHibernate.Test/Async/NHSpecificTest/GH1180/FixtureByCode.cs:127]: SQLite does not have proper decimal type, it has `TestDialect.HasBrokenDecimalType` (not sure about the name of this property but you got the idea) set to true.

File: src/NHibernate.Test/NHSpecificTest/GH1180/FixtureByCode.cs [hunk 3]
```diff
@@ -90,9 +90,9 @@ public void DecimalTypes()
 			using (var session = OpenSession())
 			using (var transaction = session.BeginTransaction())
 			{
-				session.Save(new Entity {Amount = 3.141m});
-				session.Save(new Entity {Amount = 42.131m});
-				session.Save(new Entity {Amount = 17.991m});
+				session.Save(new Entity {Amount = 3.14m});
+				session.Save(new Entity {Amount = 42.13m});
+				session.Save(new Entity {Amount = 17.99m});
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1180/FixtureByCode.cs]: I have not checked if it makes sense within this test to check if the expected scale is respected, but "fixing" SQLite this way cause the test to no more check this point.

If it was making sense, it is not what should be done. In such case, the test should instead be adapted by checking if the dialect has proper support of decimal.

Comment (bahusoid) [reply src/NHibernate.Test/NHSpecificTest/GH1180/FixtureByCode.cs]: It's not exactly related. For this test only constant projection scale check make sense for `Projections.Constant(20.123m)`. But still I restored this check too.

**Review by hazzik — APPROVED** [review]

**Pull request merged**